plugin/helpers_bots.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package plugin
+
+import (
+	"time"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/pkg/errors"
+)
+
+func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error) {
+	// Must provide a bot with a username
+	if bot == nil || len(bot.Username) < 1 {
+		return "", errors.New("EnsureBot was passed a bad bot")
```

Comment:
In general the context of an error should not be determined in the method, but rather by the caller. Secondly I would like to spit up the two checks into two separate errors.

```suggestion
	if bot == nil {
		return "", errors.New("provided bot is nil")
	}
	if len(bot.Username) = "" {
		return "", errors.New("a Username for the bot is required")
	}
```

Reply:
I don't know this is worth the mental overhead in reading multiple conditions. The error isn't a runtime thing it's something to help a programmer. I don't think we need to add complexity in maintain this by splitting this up.

Reply:
This error should help programmers to understand what they did wrong. I'm 1/5, not blocking if you don't want to add this,

Reply:
Just seems like unnecessary complexity. The programmer can figure this out just as fast ether way.